src/main/java/io/anserini/collection/C4Collection.java

```diff
@@ -0,0 +1,153 @@
+        this.path = path;
+    }
+
+    @Override
+    public FileSegment<C4Collection.Document> createFileSegment(Path p) throws IOException {
+        return new Segment(p);
+    }
+
+    public static class Segment extends FileSegment<C4Collection.Document>{
+        private MappingIterator<JsonNode> iterator; // iterator for JSON line objects
+        private JsonNode node = null;
+        private String filePath;
+        private String fileName;
+        private int count = 0;
+
+        public Segment(Path path) throws IOException {
+            super(path);
+            filePath = path.toString();
+            int fileNumStart = filePath.indexOf(".")+1;
+            fileName = filePath.substring(fileNumStart, fileNumStart + 5);
+            System.out.println(fileName);
+            if (filePath.endsWith(".gz")) { //.gz
+                InputStream stream = new GZIPInputStream(
+                        Files.newInputStream(path, StandardOpenOption.READ), BUFFER_SIZE);
+                bufferedReader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8));
+            } else { // plain text file
+                bufferedReader = new BufferedReader(new FileReader(filePath));
+            }
+            // reading as a json file
+            ObjectMapper mapper = new ObjectMapper();
+            // removes ctrl characters
+            String filtered = bufferedReader.lines().map(line -> line.replaceAll("[\\p{C}]","")).collect(Collectors.joining());
+            iterator = mapper.readerFor(JsonNode.class).readValues(filtered);
+            if (iterator.hasNext()){
+                node = iterator.next();
+            }
+        }
+
+        @Override
+        public void readNext() throws NoSuchElementException {
+            if (node == null) {
+                throw new NoSuchElementException("JsonNode is empty");
+            } else {
+                bufferedRecord = new C4Collection.Document(node, fileName, count);
+                if (iterator.hasNext()) { // if bufferedReader contains JSON line objects, we parse the next JSON into node
+                    node = iterator.next();
+                    count++;
+                } else {
+                    atEOF = true; // there is no more JSON object in the bufferedReader
+                }
+            }
+        }
+    }
+
+    public static class Document implements SourceDocument {
+        private String id;
+        private String contents;
+        private String raw;
+        private String url;
+        private long timestamp;
+
+        public Document(JsonNode json, String filename, int jsonLoc) {
+            this.raw = json.toPrettyString();
+            this.contents = json.get("text").asText();
+            // id is the filename appended by the document count
+            this.id = filename +  '-' + jsonLoc;
+            this.url = json.get("url").asText();
+            String dateTime = json.get("timestamp").asText();
```

Comment:
you probably want to do some null checking here?

Reply:
idt, for those fields they will always be there and in other json collections we never did null checking on fields.